src/Processors/QueryPlan/JoinStepLogical.cpp

```diff
@@ -40,85 +68,148 @@ namespace Setting
+        {
+            node = &actions_dag->addFunction(to_nullable, {node}, {});
+        }
+
+        if (node->result_name != original_name)
+            node = &actions_dag->addAlias(*node, std::move(original_name));
+
+        actions_after_join.push_back(node);
+        outputs.push_back(node);
+    }
+
+    if (outputs.empty())
+    {
+        auto column_type = std::make_shared<DataTypeUInt8>();
+        ColumnWithTypeAndName column(column_type->createColumnConst(1, 0), column_type, "__join_result_dummy");
+        const auto * node = &actions_dag->addColumn(std::move(column));
+        actions_after_join.push_back(node);
+        outputs.push_back(node);
+    }
 }
 
 JoinStepLogical::JoinStepLogical(
     SharedHeader left_header_,
     SharedHeader right_header_,
-    JoinInfo join_info_,
+    JoinOperator join_operator_,
     JoinExpressionActions join_expression_actions_,
-    Names required_output_columns_,
+    const NameSet & required_output_columns_,
+    const std::unordered_map<String, const ActionsDAG::Node *> & changed_types,
     bool use_nulls_,
     JoinSettings join_settings_,
     SortingStep::Settings sorting_settings_)
     : expression_actions(std::move(join_expression_actions_))
-    , join_info(std::move(join_info_))
-    , required_output_columns(std::move(required_output_columns_))
-    , use_nulls(use_nulls_)
+    , join_operator(std::move(join_operator_))
+    , join_settings(std::move(join_settings_))
+    , sorting_settings(std::move(sorting_settings_))
+{
+    if (!changed_types.empty())
+    {
+        /// FIXME: do not reorder join with `using` clause
+        optimized = true;
+    }
+    addToNullableIfNeeded(expression_actions, join_operator.kind, use_nulls_, required_output_columns_, actions_after_join, changed_types);
+    updateInputHeaders({left_header_, right_header_});
+}
+
+JoinStepLogical::JoinStepLogical(
+    const SharedHeader & left_header_,
+    const SharedHeader & right_header_,
+    JoinOperator join_operator_,
+    JoinExpressionActions join_expression_actions_,
+    std::vector<const ActionsDAG::Node *> actions_after_join_,
+    JoinSettings join_settings_,
+    SortingStep::Settings sorting_settings_)
+    : expression_actions(std::move(join_expression_actions_))
+    , join_operator(std::move(join_operator_))
     , join_settings(std::move(join_settings_))
     , sorting_settings(std::move(sorting_settings_))
 {
+    actions_after_join = std::move(actions_after_join_);
     updateInputHeaders({left_header_, right_header_});
+
+    auto current_actions_dag = expression_actions.getActionsDAG();
+    for (const auto * node_after_join : actions_after_join)
+    {
+        if (!current_actions_dag->containsNode(node_after_join))
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Node {} is not in the expression actions dag", fmt::ptr(node_after_join));
+    }
+}
+
+std::unordered_set<JoinTableSide> JoinStepLogical::typeChangingSides() const
```

Comment:
Can we please replace `std::unordered_set<JoinTableSide>` to enum with 4 values?